dagstore_async.go

```diff
@@ -43,40 +58,24 @@ func (d *DAGStore) acquireAsync(ctx context.Context, w *waiter, s *Shard, mnt mo
 func (d *DAGStore) initializeAsync(ctx context.Context, s *Shard, mnt mount.Mount) {
 	reader, err := mnt.Fetch(ctx)
 	if err != nil {
-		err = fmt.Errorf("failed to acquire reader of mount: %w", err)
-		_ = d.queueTask(&task{op: OpShardFail, shard: s, err: err}, d.completionCh)
+		_ = d.failShard(s, fmt.Errorf("failed to acquire reader of mount: %w", err), d.completionCh)
 		return
 	}
-
 	defer reader.Close()
-
-	carreader, err := car.NewReader(reader)
+	// works for both CARv1 and CARv2.
+	idx, err := car.ReadOrGenerateIndex(reader)
```

Comment:
This is a new API, and it's good for convenience. But we will not what to use it this way, because it's opaque and that harms observability and extensibility.

1. It could take 1 second if an inline index exists, or it could take minutes if this is forced to index a 64GiB CAR. And there's no way to know. Conversely, if you check the version and the header yourself, you can gather this information and expose it (through extra states like `ShardStateGeneratingIndex`, for example).
2. For better scalability, in the future we will want to control the scheduling of index generation. Imagine you plug in an external HDD with 1TiB worth of 4GiB CARs v1, each of which is a shard. With this API, you would end up indexing all of them at once, which will lead to a poor outcome due to the IO footprint of performing random access on 256 files at the same time.

Reply:
@raulk Sure, but can we make that change when we have to make that change ?

Reply:
This is simple, makes for less code in the Sharded DAG Store that we are just getting off the ground and is completely untested in the wild and our Markets tests have just started showing green after integrating the Sharded DAG Store.

Please can we write code for the future use case when the future use case arrives ?

Reply:
We already had code like this before, but I'll open an issue to track.

Reply:
https://github.com/filecoin-project/dagstore/issues/50

Reply:
This makes a lot of sense to me.